cmd/kubeadm/app/cmd/join.go

```diff
@@ -260,6 +260,7 @@ func AddJoinOtherFlags(flagSet *flag.FlagSet, cfgPath *string, ignorePreflightEr
 // Join defines struct used by kubeadm join command
 type Join struct {
 	cfg                   *kubeadmapi.JoinConfiguration
+	initCfg               *kubeadmapi.InitConfiguration
```

Comment:
having both configs in a struct called Join feels weird indeed. the struct is just used for storage in the case of initcfg, so that fetchInitConfigurationFromJoinConfiguration() is not called multiple times.
if they remain like so they should be joinCfg, initCfg. also line 299 would need some renames.

on control-plane nodes the IPVS check is conditional on kube-proxy mode already. the conditional IPVS check on worker nodes seems viable in cases of worker nodes not having the IPVS modules where an error will always be thrown.

Reply:
@timothysc what we really need here is ClusterConfiguration but as of today InitConfiguration and ClusterConfiguration are still tied together in many utilities.
Fixing the above point IMO is out scope of this PR, so I'm fine with the proposed solution with one last exception already discussed with @ereslibre